### PR TITLE
[SPARK-25430][SQL] Add map parameter for withColumnRenamed

### PR DESCRIPTION
**Pull request by goungoun**
## What changes were proposed in this pull request?
This PR allows withColumnRenamed with a map input argument

## How was this patch tested?
unit tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2300,6 +2300,37 @@ class Dataset[T] private[sql](
...
+   * This is a no-op if schema doesn't contain existingNames in columnMap.
+   * {{{
+   *   df.withColumnRenamed(Map(
+   *     "c1" -> "first_column",
+   *     "c2" -> "second_column"
+   *   ))
+   * }}}
+   *
+   * @group untypedrel
+   * @since 2.4.0
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: branch-2.4 is cut out. We will probably target 3.0.0 if we happen to add new APIs.

**Comment by HyukjinKwon** [comment]
Can we simply call the API multiple times? I think we haven't usually added such aliases for an API unless there's strong argument for it.

**Comment by goungoun** [comment]
@HyukjinKwon , thanks for your review. Actually, that is the reason that I open this pull request. I think it is better to giving reusable option to users than repeating too much of same code in their analysis. In notebook environment, whenever visualization is required in the middle of the analysis, I had to convert column names rather than using it as it is so that I can deliver right messages to the report readers. During the process, I had to repeat withColumenRenamed too many times. 

So, I've researched how the other users are trying to overcome the limitation. It seems that users tend to use foldleft or for loop with withColumnRenamed which can cause performance issue creating too many dataframes inside of Spark engine even without knowing it. The arguments can be found as follows.

StackOverflow
- https://stackoverflow.com/questions/38798567/pyspark-rename-more-than-one-column-using-withcolumnrenamed
- https://stackoverflow.com/questions/35592917/renaming-column-names-of-a-dataframe-in-spark-scala?noredirect=1&lq=1

Spark Issues
[SPARK-12225] Support adding or replacing multiple columns at once in DataFrame API

[SPARK-21582] DataFrame.withColumnRenamed cause huge performance overhead
If foldleft is used, too many columns can cause performance issue

**Comment by HyukjinKwon** [comment]
The performance issue was introduced by repeating query plan analysis, which is resolved in the current master if I am not mistaken - if you're in doubt, I would suggest to do a quick benchamrk. I think this is something we should do it with one liner helper in application side code.

**Review by gatorsmile — COMMENTED** [review]
Adding a new API is not needed especially after we improve our resolution algorithm in 2.4 release. See the commit: https://github.com/apache/spark/commit/4e861db5f149e10fd8dfe6b3c1484821a590b1e8

**Comment by goungoun** [comment]
Awesome! @HyukjinKwon , @gatorsmile thanks for good information. Let me look into it further. By the way, I still hope this conversation is open to users' voice, not limited with developers' perspective. Like me who have to do data wrangling/engineering everyday, it makes things easier.

**Pull request closed without merging**